### PR TITLE
#163422176 Feedback implementation 

### PR DESCRIPTION
**Pull request by SimonAwiti**
#### What does this PR do?
contains all the feedback suggested

#### Description of Task to be completed?
Add token-based authentication 
Add extra endpoints suggested in other branches
##### Clone and Configure a Flask Project
Login into your github account and open the project folder then follow the instruction on how to clone the existing project. It should be something similar to:
```
git clone https://github.com/SimonAwiti/Questioner-APIs.git
```
Install the requirements by typing:
pip install -r requirements.txt

## How to run the app
On the terminal type:
flask run after setting up the environment
#### Any background context you want to provide?
The technology to be used involve Python, Flask_extended, psychopg2
#### Relevant pivotal tracker stories
https://www.pivotaltracker.com/n/projects/2215276
#163422176
screenshot from 2019-01-23 23-45-48

### CODE REVIEW

File: app/API/utilities/validator.py [hunk 1]
```diff
@@ -1,64 +1,129 @@
...
+        dates = args['happeningOn']
+        try:
+            happenning = datetime.strptime(dates, "%d/%m/%Y")
+        except ValueError as exception:
+            return {}
+        days = (happenning - datetime.utcnow()).days
+        if days < 0:
+            return{
+                "status": 401,
+                "error": "Your happening date must be ahead of date of meetup posting"
```
Comment (johnmusiu) [inline app/API/utilities/validator.py]:
```suggestion
                "error": "A meetup cannot be scheduled in the past."
```

Consider changing this error message to a more understandable version as suggested

File: app/API/version1/users/validators.py [hunk 2]
```diff
@@ -4,24 +4,54 @@ def validate_data_signup(args):
     """validate user details"""
     try:
         if " " in args["password"] or " " in args["email"]:
-            return "password and email should be one word, no spaces"
+            return{
+                "status": 401,
+                "error": "password and email should be one word, no spaces"
```
Comment (johnmusiu) [inline app/API/version1/users/validators.py]: I suppose passwords can have spaces

File: app/API/version1/users/validators.py [hunk 3]
```diff
@@ -4,24 +4,54 @@ def validate_data_signup(args):
...
+                    "status": 401,
+                    "error": "Fields cannot be left empty"
+                    }, 401
+        elif not re.match(r'^(?=.*[a-z])(?=.*\d)(?=.*[A-Z])(?:.{8,})$', args["password"]):
+            return{
+                    "status": 401,
+                    "error": "Please provide a strong password"
+                    }, 401
+
         elif len(args['password'].strip()) < 5:
```
Comment (johnmusiu) [inline app/API/version1/users/validators.py:26]: passwords are allowed to have spaces, arent they?

File: app/API/version2/meetups/models.py [hunk 4]
```diff
@@ -127,16 +153,19 @@ def add_meetup(self, location, topic ):
...
             connect.commit()
+            meetup = cursor.fetchone()
             response = jsonify({'status': 201,
-                                "msg":'Meetup Successfully Created'})
+                                "msg":'Meetup Successfully Created',
+                                "data": Meetups.json(meetup)})
             response.status_code = 201
             return response
         except (Exception, psycopg2.DatabaseError) as error:
+            print(error)
```
Comment (johnmusiu) [inline app/API/version2/meetups/models.py]: Consider deleting this print statement

File: app/API/version2/questions/models.py [hunk 5]
```diff
@@ -105,19 +110,40 @@ def upvote_question(self, question_id):
...
     def downvote_question(self, question_id):
-        question = Helper.check_if_question_posted_exists(self, question_id)
+        question = Helper.check_ques(self,question_id)
+        if not question:
+            return {"message": "That question does not exist"}, 404
+        user_id = question[3]
+        votes = question[6]
+        if user_id:
+            if votes > 0:
+                return {"Message": "you can't vote again"}
```
Comment (johnmusiu) [inline app/API/version2/questions/models.py]: Should this response have a Status code?

Comment (SimonAwiti) [reply app/API/version2/questions/models.py]: Worked on all the feedback. Thanks for the feedback

**Pull request merged**